File: tests/filepath_test.cpp
#include <qk/qk_filepath.h>
#include <catch2/catch_test_macros.hpp>

using namespace qk::filepath;

TEST_CASE("Filepath clean function", "[filepath]") {
    SECTION("Basic path cleaning") {
        REQUIRE(clean("path/to/../file") == from_slash("path/file"));
        REQUIRE(clean("path/./file") == from_slash("path/file"));
        REQUIRE(
            clean(string("path") + SEPARATOR + "" + SEPARATOR + "file") == from_slash("path/file")
        );
    }

    SECTION("Rooted paths") {
        REQUIRE(clean("/path/to/../file") == from_slash("/path/file"));
        REQUIRE(clean("/.") == from_slash("/"));
        REQUIRE(clean("/..") == from_slash("/"));
    }

    SECTION("Additional edge cases") {
        REQUIRE(clean(from_slash("path\\to/file")) == from_slash("path/to/file"));
        REQUIRE(clean("a/b/../../c") == from_slash("c"));
        REQUIRE(clean("C:foo/..") == from_slash("C:."));
        REQUIRE(clean("\\\\?\\C:\\foo\\..") == from_slash("\\\\?\\C:\\"));
    }

    SECTION("Empty or dot paths") {
        REQUIRE(clean("") == ".");
        REQUIRE(clean(".") == ".");
        REQUIRE(clean("..") == "..");
    }

#ifdef _WIN32
    SECTION("Windows volume name handling") {
        REQUIRE(clean("C:\\path\\to\\..\\file") == "C:\\path\\file");
        REQUIRE(clean("C:\\") == "C:\\");
        REQUIRE(clean("\\\\server\\share\\path\\..\\file") == "\\\\server\\share\\file");
    }
#endif
}

TEST_CASE("Filepath split function", "[filepath]") {
    string dir, file;

    SECTION("Basic path splitting") {
        split(from_slash("path/to/file.txt"), &dir, &file);
        REQUIRE(dir == from_slash("path/to/"));
        REQUIRE(file == "file.txt");
    }

    SECTION("Rooted path splitting") {
        split(from_slash("/file.txt"), &dir, &file);
        REQUIRE(dir == from_slash("/"));
        REQUIRE(file == "file.txt");
    }

    SECTION("No directory") {
        split("file.txt", &dir, &file);
        REQUIRE(dir == "");
        REQUIRE(file == "file.txt");
    }

    SECTION("Edge cases") {
        split("", &dir, &file);
        REQUIRE(dir == "");
        REQUIRE(file == "");
        split(from_slash("path/"), &dir, &file);
        REQUIRE(dir == from_slash("path/"));
        REQUIRE(file == "");
        split("C:", &dir, &file);
        REQUIRE(dir == "C:");
        REQUIRE(file == "");
        split("\\\\server\\share", &dir, &file);
        REQUIRE(dir == "\\\\server\\share");
        REQUIRE(file == "");
    }

#ifdef _WIN32
    SECTION("Windows volume name") {
        split("C:\\path\\file.txt", &dir, &file);
        REQUIRE(dir == "C:\\path\\");
        REQUIRE(file == "file.txt");
    }
#endif
}

TEST_CASE("Filepath ext function", "[filepath]") {
    SECTION("Basic extension extraction") {
        REQUIRE(ext("file.txt") == ".txt");
        REQUIRE(ext("file.tar.gz") == ".gz");
        REQUIRE(ext("file") == "");
    }

    SECTION("Additional cases") {
        REQUIRE(ext(".gitignore") == ".gitignore");
        REQUIRE(ext("file..txt") == ".txt");
        REQUIRE(ext(from_slash("path/.")) == ".");
    }

    SECTION("Path with directories") { REQUIRE(ext(from_slash("path/to/file.txt")) == ".txt"); }

    SECTION("No extension") { REQUIRE(ext(from_slash("path/to/file")) == ""); }
}

TEST_CASE("Filepath base function", "[filepath]") {
    SECTION("Basic base extraction") {
        REQUIRE(base(from_slash("path/to/file.txt")) == "file.txt");
        REQUIRE(base("file.txt") == "file.txt");
        REQUIRE(base("") == ".");
    }

    SECTION("Rooted paths") {
        REQUIRE(base(from_slash("/path/to/file.txt")) == "file.txt");
#ifdef _WIN32
        REQUIRE(base("C:\\path\\to\\file.txt") == "file.txt");
#endif
    }

    SECTION("Trailing separators") { REQUIRE(base(from_slash("path/to//")) == "to"); }

#ifdef _WIN32
    SECTION("Windows edge cases") { REQUIRE(base("C:") == string({SEPARATOR})); }
#endif
}

TEST_CASE("Filepath dir function", "[filepath]") {
    SECTION("Basic directory extraction") {
        REQUIRE(dir(from_slash("path/to/file.txt")) == from_slash("path/to"));
        REQUIRE(dir("file.txt") == ".");
    }

    SECTION("Rooted paths") {
        REQUIRE(dir(from_slash("/path/to/file.txt")) == from_slash("/path/to"));
#ifdef _WIN32
        REQUIRE(dir("C:\\path\\to\\file.txt") == "C:\\path\\to");
#endif
    }

    SECTION("Edge cases") {
        REQUIRE(dir("") == ".");
        REQUIRE(dir("C:") == "C:.");
        REQUIRE(dir("\\\\server\\share\\file") == "\\\\server\\share\\");
        REQUIRE(dir(from_slash("path/")) == from_slash("path"));
    }
}

TEST_CASE("Filepath slash conversion", "[filepath]") {
    SECTION("To slash") {
        REQUIRE(to_slash(string("path") + SEPARATOR + "to" + SEPARATOR + "file") == "path/to/file");
        REQUIRE(to_slash("path/to/file") == "path/to/file");
        REQUIRE(to_slash("") == "");
        REQUIRE(to_slash(string(1, SEPARATOR)) == "/");
    }

    SECTION("From slash") {
        REQUIRE(
            from_slash("path/to/file") == string("path") + SEPARATOR + "to" + SEPARATOR + "file"
        );
        REQUIRE(from_slash("") == "");
    }
}

TEST_CASE("Filepath volume name", "[filepath]") {
    SECTION("Non-Windows volume name") { REQUIRE(volume_name(from_slash("path/to/file")) == ""); }

#ifdef _WIN32
    SECTION("Windows volume name") {
        REQUIRE(volume_name("C:\\path\\to\\file") == "C:");
        REQUIRE(volume_name("\\\\server\\share\\file") == "\\\\server\\share");
    }

    SECTION("Windows edge cases") {
        REQUIRE(volume_name("\\\\?\\C:\\file") == "\\\\?\\C:");
        REQUIRE(volume_name("\\\\server") == "\\\\server");
        REQUIRE(volume_name("C") == "");
    }
#endif
}

File: tests/filepath_test.go
package main

import (
	"path/filepath"
	"runtime"
	"strings"
	"testing"
)

func TestFilepathClean(t *testing.T) {
	t.Run("Basic path cleaning", func(t *testing.T) {
		tests := []struct {
			input    string
			expected string
		}{
			{"path/to/../file", "path/file"},
			{"path/./file", "path/file"},
			{filepath.Join("path", "", "file"), "path/file"},
		}
		for _, tt := range tests {
			t.Run(tt.input, func(t *testing.T) {
				got := filepath.Clean(tt.input)
				expected := filepath.FromSlash(tt.expected)
				if got != expected {
					t.Errorf("Clean(%q) = %q; want %q", tt.input, got, expected)
				}
			})
		}
	})

	t.Run("Rooted paths", func(t *testing.T) {
		tests := []struct {
			input    string
			expected string
		}{
			{"/path/to/../file", "/path/file"},
			{"/.", "/"},
			{"/..", "/"},
		}
		for _, tt := range tests {
			t.Run(tt.input, func(t *testing.T) {
				got := filepath.Clean(tt.input)
				expected := filepath.FromSlash(tt.expected)
				if got != expected {
					t.Errorf("Clean(%q) = %q; want %q", tt.input, got, expected)
				}
			})
		}
	})

	t.Run("Additional edge cases", func(t *testing.T) {
		tests := []struct {
			input    string
			expected string
		}{
			{filepath.FromSlash("path\\to/file"), "path/to/file"},
			{"a/b/../../c", "c"},
			{"C:foo/..", "C:."},
			{"\\\\?\\C:\\foo\\..", "\\\\?\\C:\\"},
		}
		for _, tt := range tests {
			t.Run(tt.input, func(t *testing.T) {
				got := filepath.Clean(tt.input)
				expected := tt.expected
				if runtime.GOOS == "windows" {
					expected = strings.ReplaceAll(tt.expected, "/", "\\")
				} else {
					expected = filepath.FromSlash(tt.expected)
				}
				if got != expected {
					t.Errorf("Clean(%q) = %q; want %q", tt.input, got, expected)
				}
			})
		}
	})

	t.Run("Empty or dot paths", func(t *testing.T) {
		tests := []struct {
			input    string
			expected string
		}{
			{"", "."},
			{".", "."},
			{"..", ".."},
		}
		for _, tt := range tests {
			t.Run(tt.input, func(t *testing.T) {
				got := filepath.Clean(tt.input)
				if got != tt.expected {
					t.Errorf("Clean(%q) = %q; want %q", tt.input, got, tt.expected)
				}
			})
		}
	})

	if runtime.GOOS == "windows" {
		t.Run("Windows volume name handling", func(t *testing.T) {
			tests := []struct {
				input    string
				expected string
			}{
				{"C:\\path\\to\\..\\file", "C:\\path\\file"},
				{"C:\\", "C:\\"},
				{"\\\\server\\share\\path\\..\\file", "\\\\server\\share\\file"},
			}
			for _, tt := range tests {
				t.Run(tt.input, func(t *testing.T) {
					got := filepath.Clean(tt.input)
					if got != tt.expected {
						t.Errorf("Clean(%q) = %q; want %q", tt.input, got, tt.expected)
					}
				})
			}
		})
	}
}

func TestFilepathSplit(t *testing.T) {
	t.Run("Basic path splitting", func(t *testing.T) {
		dir, file := filepath.Split(filepath.FromSlash("path/to/file.txt"))
		wantDir := filepath.FromSlash("path/to/")
		wantFile := "file.txt"
		if dir != wantDir || file != wantFile {
			t.Errorf("Split(%q) = (%q, %q); want (%q, %q)", "path/to/file.txt", dir, file, wantDir, wantFile)
		}
	})

	t.Run("Rooted path splitting", func(t *testing.T) {
		dir, file := filepath.Split(filepath.FromSlash("/file.txt"))
		wantDir := filepath.FromSlash("/")
		wantFile := "file.txt"
		if dir != wantDir || file != wantFile {
			t.Errorf("Split(%q) = (%q, %q); want (%q, %q)", "/file.txt", dir, file, wantDir, wantFile)
		}
	})

	t.Run("No directory", func(t *testing.T) {
		dir, file := filepath.Split("file.txt")
		wantDir := ""
		wantFile := "file.txt"
		if dir != wantDir || file != wantFile {
			t.Errorf("Split(%q) = (%q, %q); want (%q, %q)", "file.txt", dir, file, wantDir, wantFile)
		}
	})

	t.Run("Edge cases", func(t *testing.T) {
		tests := []struct {
			input    string
			wantDir  string
			wantFile string
		}{
			{"", "", ""},
			{filepath.FromSlash("path/"), filepath.FromSlash("path/"), ""},
			{"C:", "C:", ""},
			{"\\\\server\\share", "\\\\server\\share", ""},
		}
		for _, tt := range tests {
			t.Run(tt.input, func(t *testing.T) {
				dir, file := filepath.Split(tt.input)
				wantDir := tt.wantDir
				if runtime.GOOS == "windows" {
					wantDir = strings.ReplaceAll(tt.wantDir, "/", "\\")
				}
				if dir != wantDir || file != tt.wantFile {
					t.Errorf("Split(%q) = (%q, %q); want (%q, %q)", tt.input, dir, file, wantDir, tt.wantFile)
				}
			})
		}
	})

	if runtime.GOOS == "windows" {
		t.Run("Windows volume name", func(t *testing.T) {
			dir, file := filepath.Split("C:\\path\\file.txt")
			wantDir := "C:\\path\\"
			wantFile := "file.txt"
			if dir != wantDir || file != wantFile {
				t.Errorf("Split(%q) = (%q, %q); want (%q, %q)", "C:\\path\\file.txt", dir, file, wantDir, wantFile)
			}
		})
	}
}

func TestFilepathExt(t *testing.T) {
	t.Run("Basic extension extraction", func(t *testing.T) {
		tests := []struct {
			input    string
			expected string
		}{
			{"file.txt", ".txt"},
			{"file.tar.gz", ".gz"},
			{"file", ""},
		}
		for _, tt := range tests {
			t.Run(tt.input, func(t *testing.T) {
				got := filepath.Ext(tt.input)
				if got != tt.expected {
					t.Errorf("Ext(%q) = %q; want %q", tt.input, got, tt.expected)
				}
			})
		}
	})

	t.Run("Additional cases", func(t *testing.T) {
		tests := []struct {
			input    string
			expected string
		}{
			{".gitignore", ".gitignore"},
			{"file..txt", ".txt"},
			{filepath.FromSlash("path/."), "."},
		}
		for _, tt := range tests {
			t.Run(tt.input, func(t *testing.T) {
				got := filepath.Ext(tt.input)
				if got != tt.expected {
					t.Errorf("Ext(%q) = %q; want %q", tt.input, got, tt.expected)
				}
			})
		}
	})

	t.Run("Path with directories", func(t *testing.T) {
		got := filepath.Ext(filepath.FromSlash("path/to/file.txt"))
		want := ".txt"
		if got != want {
			t.Errorf("Ext(%q) = %q; want %q", "path/to/file.txt", got, want)
		}
	})

	t.Run("No extension", func(t *testing.T) {
		got := filepath.Ext(filepath.FromSlash("path/to/file"))
		want := ""
		if got != want {
			t.Errorf("Ext(%q) = %q; want %q", "path/to/file", got, want)
		}
	})
}

func TestFilepathBase(t *testing.T) {
	t.Run("Basic base extraction", func(t *testing.T) {
		tests := []struct {
			input    string
			expected string
		}{
			{filepath.FromSlash("path/to/file.txt"), "file.txt"},
			{"file.txt", "file.txt"},
			{"", "."},
		}
		for _, tt := range tests {
			t.Run(tt.input, func(t *testing.T) {
				got := filepath.Base(tt.input)
				if got != tt.expected {
					t.Errorf("Base(%q) = %q; want %q", tt.input, got, tt.expected)
				}
			})
		}
	})

	t.Run("Rooted paths", func(t *testing.T) {
		tests := []struct {
			input    string
			expected string
		}{
			{filepath.FromSlash("/path/to/file.txt"), "file.txt"},
		}
		if runtime.GOOS == "windows" {
			tests = append(tests, struct {
				input    string
				expected string
			}{"C:\\path\\to\\file.txt", "file.txt"})
		}
		for _, tt := range tests {
			t.Run(tt.input, func(t *testing.T) {
				got := filepath.Base(tt.input)
				if got != tt.expected {
					t.Errorf("Base(%q) = %q; want %q", tt.input, got, tt.expected)
				}
			})
		}
	})

	t.Run("Trailing separators", func(t *testing.T) {
		got := filepath.Base(filepath.FromSlash("path/to//"))
		expected := "to"
		if got != expected {
			t.Errorf("Base(%q) = %q; want %q", "path/to//", got, expected)
		}
	})

	if runtime.GOOS == "windows" {
		t.Run("Windows edge cases", func(t *testing.T) {
			got := filepath.Base("C:")
			expected := string(filepath.Separator)
			if got != expected {
				t.Errorf("Base(%q) = %q; want %q", "C:", got, expected)
			}
		})
	}
}

func TestFilepathDir(t *testing.T) {
	t.Run("Basic directory extraction", func(t *testing.T) {
		tests := []struct {
			input    string
			expected string
		}{
			{filepath.FromSlash("path/to/file.txt"), "path/to"},
			{"file.txt", "."},
		}
		for _, tt := range tests {
			t.Run(tt.input, func(t *testing.T) {
				got := filepath.Dir(tt.input)
				expected := filepath.FromSlash(tt.expected)
				if got != expected {
					t.Errorf("Dir(%q) = %q; want %q", tt.input, got, expected)
				}
			})
		}
	})

	t.Run("Rooted paths", func(t *testing.T) {
		tests := []struct {
			input    string
			expected string
		}{
			{filepath.FromSlash("/path/to/file.txt"), "/path/to"},
		}
		if runtime.GOOS == "windows" {
			tests = append(tests, struct {
				input    string
				expected string
			}{"C:\\path\\to\\file.txt", "C:\\path\\to"})
		}
		for _, tt := range tests {
			t.Run(tt.input, func(t *testing.T) {
				got := filepath.Dir(tt.input)
				expected := tt.expected
				if runtime.GOOS == "windows" {
					expected = strings.ReplaceAll(expected, "/", "\\")
				} else {
					expected = filepath.FromSlash(expected)
				}
				if got != expected {
					t.Errorf("Dir(%q) = %q; want %q", tt.input, got, expected)
				}
			})
		}
	})

	t.Run("Edge cases", func(t *testing.T) {
		tests := []struct {
			input    string
			expected string
		}{
			{"", "."},
			{"C:", "C:."},
			{"\\\\server\\share\\file", "\\\\server\\share\\"},
			{filepath.FromSlash("path/"), "path"},
		}
		for _, tt := range tests {
			t.Run(tt.input, func(t *testing.T) {
				got := filepath.Dir(tt.input)
				expected := tt.expected
				if runtime.GOOS == "windows" {
					expected = strings.ReplaceAll(expected, "/", "\\")
				} else {
					expected = filepath.FromSlash(expected)
				}
				if got != expected {
					t.Errorf("Dir(%q) = %q; want %q", tt.input, got, expected)
				}
			})
		}
	})
}

func TestFilepathSlashConversion(t *testing.T) {
	t.Run("To slash", func(t *testing.T) {
		tests := []struct {
			input    string
			expected string
		}{
			{filepath.Join("path", "to", "file"), "path/to/file"},
			{"path/to/file", "path/to/file"},
			{"", ""},
			{string(filepath.Separator), "/"},
		}
		for _, tt := range tests {
			t.Run(tt.input, func(t *testing.T) {
				got := filepath.ToSlash(tt.input)
				if got != tt.expected {
					t.Errorf("ToSlash(%q) = %q; want %q", tt.input, got, tt.expected)
				}
			})
		}
	})

	t.Run("From slash", func(t *testing.T) {
		tests := []struct {
			input    string
			expected string
		}{
			{"path/to/file", filepath.Join("path", "to", "file")},
			{"", ""},
		}
		for _, tt := range tests {
			t.Run(tt.input, func(t *testing.T) {
				got := filepath.FromSlash(tt.input)
				if got != tt.expected {
					t.Errorf("FromSlash(%q) = %q; want %q", tt.input, got, tt.expected)
				}
			})
		}
	})

	t.Run("Edge cases", func(t *testing.T) {
		// Already covered in above
	})
}

func TestFilepathVolumeName(t *testing.T) {
	t.Run("Non-Windows volume name", func(t *testing.T) {
		got := filepath.VolumeName(filepath.FromSlash("path/to/file"))
		want := ""
		if got != want {
			t.Errorf("VolumeName(%q) = %q; want %q", "path/to/file", got, want)
		}
	})

	if runtime.GOOS == "windows" {
		t.Run("Windows volume name", func(t *testing.T) {
			tests := []struct {
				input    string
				expected string
			}{
				{"C:\\path\\to\\file", "C:"},
				{"\\\\server\\share\\file", "\\\\server\\share"},
			}
			for _, tt := range tests {
				t.Run(tt.input, func(t *testing.T) {
					got := filepath.VolumeName(tt.input)
					if got != tt.expected {
						t.Errorf("VolumeName(%q) = %q; want %q", tt.input, got, tt.expected)
					}
				})
			}
		})

		t.Run("Windows edge cases", func(t *testing.T) {
			tests := []struct {
				input    string
				expected string
			}{
				{"\\\\?\\C:\\file", "\\\\?\\C:"},
				{"\\\\server", "\\\\server"},
				{"C", ""},
			}
			for _, tt := range tests {
				t.Run(tt.input, func(t *testing.T) {
					got := filepath.VolumeName(tt.input)
					if got != tt.expected {
						t.Errorf("VolumeName(%q) = %q; want %q", tt.input, got, tt.expected)
					}
				})
			}
		})
	}
}
